src/test/java/com/johncnstn/auth/constraints/PasswordValidatorTest.java

```diff
@@ -0,0 +1,57 @@
+package com.johncnstn.auth.constraints;
+
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class PasswordValidatorTest {
+
+    @Test
+    public void testNull() {
+        testPassword(null, false);
+    }
+
+    @Test
+    public void testEmpty() {
+        testPassword("", false);
+    }
+
+    @Test
+    public void testWithWhitespace() {
+        testPassword("Password 1", false);
+    }
+
+    @Test
+    public void testWithNonAscii() {
+        testPassword("Passwordя1", false);
+    }
+
+    @Test
+    public void testWithoutAnyLower() {
+        testPassword("PASSWORD1", false);
+    }
+
+    @Test
+    public void testWithoutAnyUpper() {
+        testPassword("password1", false);
+    }
+
+    @Test
+    public void testWithoutDigits() {
+        testPassword("password", false);
+    }
+
+    @Test
+    public void testFine() {
+        testPassword("Password1", true);
+    }
+
+    @Test
+    private void testPassword(String password, boolean expected) {
+        // WHEN
+        var result = PasswordValidator.isValid(password);
```

Comment:
PasswordValidator import, please.

Reply:
I feel that here, we should not do a static import of PasswordValidator.isValid as it makes less clear about what we are actually testing.

Reply:
We don't need info from where we are using this method.
Just isValid and password as a variable would be sufficient.

Reply:
Here, we actually need this info. I agree that it's not needed when we are using this method in code. In code, we just need to state that we want to validate a password, not depending on where we are taking the validation routine, abstracting away the actual validation logic owner. In a test method, we want to say that we are testing a concrete method of a concrete class. If you still feel that we have to omit the class here, I will do it as it is a minor thing.

Reply:
As we discussed earlier, let's name methods like isPasswordValid for static methods.
If it's @Component - then just isValid.